docs: add design doc for Ceph COSI driver (#11995)
rook/rook: design/ceph/object/ceph-cosi-driver.md

## Integration plan with Rook

The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.

### Pre-requisites

- COSI CRDs should be installed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
```

- COSI controller should be deployed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
```

### End to End Workflow

#### Changes in Common.yaml

Following contents need to be append to `common.yaml` :

- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>

> #### cephcosi.ceph.rook.io CRD

Reviewer: After discussion today:

COSI's intent is one driver per k8s cluster, so let's keep to that. 

Having a default behavior in Rook to deploy Ceph COSI driver if any CephObjectStore is deployed is a good one. Users shouldn't need to set any COSI config by default, just like with CSI today.

CRD still gives us some benefit in having a typed API that helps users with config and config validation compared to ConfigMap. Let's use CRD.

CRD can be deployed optionally if users want to override Rook's built-in defaults for COSI deployment. For example, for overriding the container image, labels, resource requests/limits, tolerations, etc. on Ceph COSI pods.

I had a thought that if we are using the CRD primarily for configuration rather than control of whether the thing exists, it might be good to call it `CephCOSIConfig` rather than `CephCOSI`. @travisn ? The CRD may also benefit from a `deploymentStrategy` option that has options `AlwaysDeploy`, `NeverDeploy`, and `AutoDeploy` where `AutoDeploy` is Rook's internal default to deploy when a CephObjectStore exists.

> #### cephcosi.ceph.rook.io CRD

Reviewer 2: > COSI's intent is one driver per k8s cluster, so let's keep to that.

Agreed, let's go with one driver per Rook operator, of which we expect one.
 
> Having a default behavior in Rook to deploy Ceph COSI driver if any CephObjectStore is deployed is a good one. Users shouldn't need to set any COSI config by default, just like with CSI today.
> 
> CRD still gives us some benefit in having a typed API that helps users with config and config validation compared to ConfigMap. Let's use CRD.

Agreed on the CRD since it's scoped with the operator, instead of with the CephCluster

> CRD can be deployed optionally if users want to override Rook's built-in defaults for COSI deployment. For example, for overriding the container image, labels, resource requests/limits, tolerations, etc. on Ceph COSI pods.

> I had a thought that if we are using the CRD primarily for configuration rather than control of whether the thing exists, it might be good to call it `CephCOSIConfig` rather than `CephCOSI`. @travisn ? 

IMO CRDs are implicitly about config, so adding the `Config` suffix seems unnecessary to me. What about `CephCOSIDriver`? 

> The CRD may also benefit from a `deploymentStrategy` option that has options `AlwaysDeploy`, `NeverDeploy`, and `AutoDeploy` where `AutoDeploy` is Rook's internal default to deploy when a CephObjectStore exists.

What is the difference between `AlwaysDeploy` and `AutoDeploy`? They seem the same to me at least in the case of the COSI driver. If we don't have a difference, I'd vote just for a boolean which simply allows disabling the driver.

> #### cephcosi.ceph.rook.io CRD

Author: Auto --> deploys when first object store is created
Force --> will create when CRD is created, do not wait for object store
Disable --> disbales ceph cosi driver

> #### cephcosi.ceph.rook.io CRD

Reviewer 2: I see now, the main difference is `Force` allows creating the driver before the first object store is created.

The users need to define following CRD to bring up the COSI driver for the ceph object store.

```yaml
apiVersion: ceph.rook.io/v1
kind: CephCOSI
metadata:
  name: rook-ceph-cosi
  namespace: rook-ceph
spec:
  objectStoreName: <object-store-name>
```

> The current COSI architecture allows driver to interact with one object store endpoint with limitations in the APIs. Hence planning to support one object store for the driver in Rook 1.12. When multiple object stores are supported in the future, the user can define `spec.objectStoreName` to `all`. Rook Operator allows one such CRD in the cluster aka one ceph cosi driver for the cluster. In future this driver can support multiple object stores.

Reviewer 2: Why do we need `spec.objectStoreName: all`? Is that a setting in the ceph COSI driver? Can we make that the default instead of requiring the user to specify it?

> 

Reviewer 2: The design doc needs a section on what details we need to persist to map the buckets from their COSI ID to their object store, etc. We need to be clear where that metadata is stored and what metadata is needed

```bash
kubectl create -f deploy/examples/cosi/bucketclass.yaml
kubectl create -f deploy/examples/cosi/bucketclaim.yaml
kubectl create -f deploy/examples/cosi/bucketaccessclass.yaml
kubectl create -f deploy/examples/cosi/bucketaccess.yaml
```

#### Consuming COSI Bucket

The user needs to mount the secret as volume created by `BucketAccess` to the application pod. The user can use the secret to access the bucket by parsing the mounted file.

```yaml
spec:
  containers:
      volumeMounts:
        - name: cosi-secrets
          mountPath: /data/cosi
  volumes:
  - name: cosi-secrets
    secret:
      secretName: ba-secret
```

```bash
# cat /data/cosi/bucket_info.json

```json
    {
      apiVersion: "v1alpha1",
      kind: "BucketInfo",
      metadata: {
          name: "ba-$uuid"
      },
      spec: {
          bucketName: "ba-$uuid",
          authenticationType: "KEY",
          endpoint: "https://rook-ceph-my-store:443",
          accessKeyID: "AKIAIOSFODNN7EXAMPLE",
          accessSecretKey: "wJalrXUtnFEMI/K...",
          region: "us-east-1",
          protocols: [
            "s3"
          ]
      }
    }
```

#### Coexistence of COSI and lib-bucket-provisioner

> Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can coexist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The lib-bucket-provisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.

Reviewer: Conflicts do exist. I mentioned this in a past review.

> Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can coexist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The lib-bucket-provisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.

Reviewer: Conflicts do exist as I mentioned in a previous review

> #### Transition from libbucket provisioner to COSI

Reviewer: ```suggestion
#### Transition from lib-bucket-provisioner to COSI
```

This applied to OBC with reclaim policy is `Retain` otherwise the bucket will be deleted when OBC is deleted. So no point in migrating the OBC with `Delete` reclaim policy.

- First the user need to create a **COSI Bucket resource** pointing to the backend bucket.
- Then user can create BucketAccessClass and BucketAccess using the COSI Bucket CRD.
- Now the update application's credentials with BucketAccess secret, for OBC it was combination of secret and config map with keys word like AccessKey, SecretKey, Bucket, BucketHost etc. Here details in as JSON format in the [secret](#Consuming-COSI-Bucket).
- Finally the user need to delete the existing OBC.

### Points to remember

#### Ceph COSI Driver Requirements

- CephObjectStore should be deployed and running in the cluster

> - The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions

Reviewer: I think Rook can automatically create a `rook-ceph-object-store-<storename>-cosi-admin` user and store those credentials in a secret named `cosi-<storename>`.

BucketClasses can choose the object store used as a combo of namespace and storename, which directly translates to reading creds from `cosi-<storename>`. Optionally, to support manual deployment, this can also be taken to be namespace/<secretname>, and COSI can internally look for a secret called `<namespace>/<secret>` or `<namespace>/cosi-<secret>` to easily support Rook or user-defined.

Unfortunately, it's not possible to use globs in `resourceNames` in RBAC, but I think the benefits of keeping one secret per objectstore outweigh the privilege issues. Ultimately, the privilege problem is one present in k8s itself as pointed out by a user in the 
(discussion here)[https://github.com/kubernetes/kubernetes/issues/56582]. RBAC will have to allow Ceph COSI to read all secrets in the operator namespace.

Reviewer 2: > RBAC will have to allow Ceph COSI to read all secrets in the operator namespace.

The secrets are in the same namespace as the object stores, right? So if the object stores are in a different namespace from the operator, we would need a binding to the secrets in each namespace where object stores are supported. In any case, agreed that the benefits of a single secret per object store outweights creating a single secret.

Author: Yes

Author: @BlaineEXE @travisn : I am trying to use the same secret generated by cephobjectstoreuser. The downside is the secret name will be in this format: `rook-ceph-object-user-<store-name>-cosi` But this controller needs to handle creating the secret/user etc since we already have in ceph-object-user-controller. What do you think about it?

Author: @travisn I understand cosi-<store-name> make more sense. But `CephObjectStoreUser` does not have a custom secret name option. I am trying to reuse existing code to make this controller simpler, otherwise we need to add copy logic from the ceph-object-user-controller to here. Another drawback of perms of the backend user will be exposed vis CRDs and it can be modified via CRDs.

Reviewer 2: Sure, we can keep the pattern if it makes the implementation simpler

Reviewer: I think this is the only detail that remains worth discussing, but it is small. We can leave it for a future change even after alpha.

- The COSI controller should be deployed in the cluster

> - Rook can able to manage multiple ceph cosi drivers

Reviewer: Rook will deploy one Ceph COSI driver per k8s cluster in the same manner as Ceph CSI

# Ceph COSI Driver Support

## Targeted for v1.12

## Background

Container Object Storage Interface (COSI) is a specification for container orchestration frameworks to manage object storage. Even though there is no standard procotol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. COSI released v1alpha1 with Kubernetes 1.25. COSI is a new project and is not yet fully integrated by Kubernetes.
More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)

> It is projected that COSI will be the only supported object storage driver in the near future. In-tree drivers such as Ceph RGW will be replaced with their respective COSI drivers.

Reviewer 2: What in-tree driver would be replaced? Do you mean the OBCs? Anyway, I don't think we will be able to deprecate OBCs for a while unless we have a migration path for them to move to COSI.

Reviewer: Agreed. I flagged this same thing in my prior review: https://github.com/rook/rook/pull/11995#discussion_r1180625624

## Integration plan with Rook

The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.

### Pre-requisites

- COSI CRDs should be installed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
```

- COSI controller should be deployed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
```

- Create Service Account and RBAC for COSI driver

```bash
kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml

> kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml

Reviewer 2: I'd like to consider appending these to rook's common.yaml, then we don't need those steps. The COSI driver is expected to become common, so it's nice to just include the rbac.

```

### End to End Workflow

#### Changes in Operator.yaml

> Below option validates whether the COSI controller is deployed in the cluster. If it is not deployed, it will log an warning message and when user tries to create object store with cosi, it will reconcile until the controller is deployed.

Reviewer 2: I vote not to have this setting. The operator could also log if it finds the COSI controller, if they never use it. And if the COSI controller isn't installed, we can just document how to troubleshoot it when they see other failures.

```yaml
  env:
  - name: COSI_CONTROLLER_ENABLED
    value: "true"
```

#### Changes in CephObjectStore CRD

Following option will bring up the Ceph COSI driver for the RGW server.

```yaml

>   cephCosiDrvierSpec:

Reviewer 2: Everything in Rook is for Ceph, so I don't think we need that prefix.
```suggestion
  cosiDriver:
```

>     enabled: true

Reviewer 2: Do we really need this setting? How about if we always create the COSI driver if we detect the COSI controller is running? If they don't want the driver running, just don't install the COSI controller.

```

> By setting this option, Rook will deploy Ceph COSI driver with sidecar after creating the Ceph object store and will also create `CephObjectStoreUser` objectstorage-provisioner-<object-store-name> which provides credentials for Ceph COSI driver. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster which already present with current codebase. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for multiple ceph object stores in the cluster. Rook also need to support update of COSI related images independently.

Reviewer 2: > Rook will deploy Ceph COSI driver with sidecar 

Is there not a dedicated COSI pod? Or what pod is the driver a sidecar of?

#### Updating existing CephObjectStore Deployment

> If the user wants to enable the COSI driver for existing ceph object store, the user need to update the `cephCosiDrvierSpec` option in the CephObjectStore CRD. Rook will bring up the COSI driver for the existing ceph object store.

Reviewer 2: No need for this if we just detect the cosi controller automatically

Reviewer: I also think that there's no need to detect if the COSI controller is running. If the object store just always creates the secret/configmap resources that the ceph-cosi controller will look for to understand when a CephObjectSgtore is present, then COSI will be able to use said object store for creating buckets without anything new needing to be added to the API.

#### Creating COSI Related CRDs

> There are five different kubernetes resources related to COSI. They are `Bucket`, `BucketAccess`, `BucketAccessClass`, `BucketClass` and `BucketClaim`. The user can create these resources using following `kubectl` command in following order. All the examples will be added to deploy/examples/cosi directory in the Rook repository.

Reviewer 2: Order of creation doesn't matter, right?
```suggestion
There are five different kubernetes resources related to COSI. They are `Bucket`, `BucketAccess`, `BucketAccessClass`, `BucketClass` and `BucketClaim`. The user can create these resources using following `kubectl` command. All the examples will be added to deploy/examples/cosi directory in the Rook repository.
```

```bash
kubectl create -f deploy/examples/cosi/bucketclass.yaml
kubectl create -f deploy/examples/cosi/bucketclaim.yaml
kubectl create -f deploy/examples/cosi/bucketaccessclass.yaml
kubectl create -f deploy/examples/cosi/bucketaccess.yaml
```

#### Consuming COSI Bucket

The user needs to mount the secret as volume created by `BucketAccess` to the application pod. The user can use the secret to access the bucket by parsing the mounted file.

```yaml
spec:
  containers:
      volumeMounts:
        - name: cosi-secrets
          mountPath: /data/cosi
  volumes:
  - name: cosi-secrets
    secret:
      secretName: ba-secret
```

```bash
# cat /data/cosi/bucket_info.json

>     |-----------------------------------------------|

Reviewer 2: Better for color formatting to use the json formatter for this section
```suggestion
```json
```

    | {                                             |
    |   apiVersion: "v1alpha1",                     |
    |   kind: "BucketInfo",                         |
    |   metadata: {                                 |
    |       name: "ba-$uuid"                        |
    |   },                                          |
    |   spec: {                                     |
    |       bucketName: "ba-$uuid",                 |
    |       authenticationType: "KEY",              |
    |       endpoint: "https://s3.amazonaws.com",   |
    |       accessKeyID: "AKIAIOSFODNN7EXAMPLE",    |
    |       accessSecretKey: "wJalrXUtnFEMI/K...",  |
    |       region: "us-west-1",                    |
    |       protocols: [                            |
    |         "s3"                                  |
    |       ]                                       |
    |   }                                           |
    | }                                             |
    |-----------------------------------------------|

```

#### Coexistence of COSI and libbucket provisioner

> Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can co exist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The libbucket probvisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.

Author: Thats true, I don't know whether users can modify their policy in future. It need handled properly from the user end IMO

#### Transition from libbucket provisioner to COSI

> This applied to OBC with reclaim policy is `Retain`. First the user need to create a COSI bucket pointing to the backend bucket. Then user can create `bucket access class` and `bucket access` using the `COSI bucket`. Now the update applications credentials with bucket access credentials, finally the user need to delete the OBC.

Reviewer 2: When the OBC is deleted, is there a secret that will also be deleted? So their application would be broken unless they also update their application to read from the COSI file `/data/cosi/bucket_info.json`, correct?

### Points to remember

#### Ceph COSI Driver Requirements

> - Ceph Object Store should be deployed and running in the cluster

Reviewer: ```suggestion
- CephObjectStore should be deployed and running in the cluster
```

> - The credentials/endpoint for the ceph object store should be available by creating ceph object store user with proper permissions

Reviewer: ```suggestion
- The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions
```

- The COSI controller should be deployed in the cluster
- Rook can able to manage multiple ceph cosi drivers

> - Rook should not modify cosi resource like Bucket, BucketAccess, BucketClaim, BucketAccessClass etc.

Reviewer: ```suggestion
- Rook should not modify COSI resources like Bucket, BucketAccess, BucketAccessClass, or BucketClass.
```

# Ceph COSI Driver Support

## Targeted for v1.12

## Background

> Container Object Storage Interface (COSI) is a specification for container orchestration frameworks to manage object storage. Even though there is no standard protocol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. COSI released v1alpha1 with Kubernetes 1.25. COSI is a new project and is not yet fully integrated by Kubernetes.

Reviewer 2: > Even though there is no standard protocol defined for Object Store, it has flexibility to add support for all. 

Not sure what this means, can you reword it?

> COSI is a new project and is not yet fully integrated by Kubernetes.

Is this an old statement? It's fully implemented now for K8s, right?
```

Author: It is alpha status and it is not integrated with K8s like CSI

> More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)

Reviewer 2: ```suggestion
More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/).
```

It is projected that COSI will be the only supported object storage driver in the future and OBCs will be deprecated after some time.

## COSI Driver Deployment

> The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace.

Reviewer 2: ```suggestion
The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. It is not deployed by Rook.
```

> The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver is deployed in the same namespace as the Rook operator. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:

Reviewer 2: Can we use a deployment? From previous comments I don't understand why it needs to be a statefulset since it is stateless.
```suggestion
The Ceph COSI driver is deployed as a deployment with a [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver is deployed in the same namespace as the Rook operator. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
```

## Integration plan with Rook

The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There will be one COSI Driver support from Rook. The driver will be started automatically with default settings when first Object Store gets created. The driver will be stopped when last Object Store is deleted. The driver will be deployed in the same namespace as Rook operator. The user can provide additional setting via new `CephCOSIDriver` CRD which is owned by Rook.

### Pre-requisites

- COSI CRDs should be installed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
```

- COSI controller should be deployed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
```

### End to End Workflow

#### Changes in Common.yaml

Following contents need to be append to `common.yaml` :

- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>

#### ceph-object-cosi-controller

> The `ceph-object-cosi-controller` will span Ceph COSI Driver pod with default settings on the Rook Operator Namespace. The controller will bring up the Ceph COSI driver when first object store is created and will stop the COSI driver when last object store is deleted only if it detect COSI Controller is running on default namespace. The controller will also watch for CephCosiDriver CRD, if defined the driver will be started with the settings provided in the CRD. If the Ceph COSI driver if up and running,it  will also create `CephObjectStoreUser` named  `cosi` for each object store which internally creates a secret rook-ceph-object-user-<objectstore-name>-cosi provides credentials for the object store. This can be specified in the BucketClass and BucketAccessClass. Also this controller ensures maximum one CephCosiDriver CRD exists in the cluster. For v1.12 supports object store belong to RGW multisite are not supported.

Reviewer 2: > The `ceph-object-cosi-controller` will span Ceph COSI Driver pod with default settings

What does it mean to span the pod?

#### cephcosidriver.ceph.rook.io CRD

The users can define following CRD so that configuration related the Ceph COSI driver can be passed. This is not mandatory and the driver will be started with default settings if this CRD is not defined.

```yaml
apiVersion: ceph.rook.io/v1
kind: CephCosiDriver
metadata:
  name: rook-ceph-cosi
  namespace: rook-ceph

> spec:

Reviewer 2: We need these in the spec:
- `placement`: for node affinity, tolerations, and so on similar to the CephCluster and other CRs
- `resources`: for memory/cpu

  deploymentStrategy: "Auto"
```

> The `deploymentStrategy` can be `Auto` or `Force` or `Disable`. If `Auto` is specified the driver will be started automatically when first object store is created and will be stopped when last object store is deleted. If `Force` is specified the driver will be started when the CRD is created and will be stopped when the CRD is deleted. The user can also disable the driver by specifying `Disable` as the value.

Reviewer 3: Is there really any value in the "Disable" option? What use-case is there? Further, is there really much value in not deploying the driver at all? Seems strange to deploy the driver and then not use it and may be more work than it's worth.

Author: From 1.12, the cosi driver will be up by default when the object store is created and running like CSI. But for CSI we have `operator.yaml` to disable it. The intention is the same, since the COSI driver relies on CRD more than the Operator Configs this option is introduced
@travisn @BlaineEXE ^^

#### Creating COSI Related CRDs

There are five different kubernetes resources related to COSI. They are `Bucket`, `BucketAccess`, `BucketAccessClass`, `BucketClass` and `BucketClaim`. The user can create these resources using following `kubectl` command. All the examples will be added to deploy/examples/cosi directory in the Rook repository.

```bash
kubectl create -f deploy/examples/cosi/bucketclass.yaml
kubectl create -f deploy/examples/cosi/bucketclaim.yaml
kubectl create -f deploy/examples/cosi/bucketaccessclass.yaml
kubectl create -f deploy/examples/cosi/bucketaccess.yaml
```

User need refer Secret which contains access credentials for the object store in the `Parameter`  field of `BucketClass` and `BucketAccessClass` CRD as below:

```yaml
Parameter
  credentialsReference: "rook-ceph-object-user-<objectstore-name>-cosi"

>   namespaceReference: "<objectstore-store-namespace>"

Reviewer 2: Do we have control of these parameter names? If so, what about `objectStoreSecretName` and `objectStoreNamespace`?

Author: Yes we have, but I was trying for a more generic hence didn't append `objectStore`

Reviewer 2: We could put more than just the credentials in the secret, right? So it seems helpful to have the descriptive name so they know what secret and namespace should be used. Perhaps `objectUserSecretName` and `objectStoreNamespace`

```

#### Consuming COSI Bucket

The user needs to mount the secret as volume created by `BucketAccess` to the application pod. The user can use the secret to access the bucket by parsing the mounted file.

```yaml
spec:
  containers:
      volumeMounts:
        - name: cosi-secrets
          mountPath: /data/cosi
  volumes:
  - name: cosi-secrets
    secret:
      secretName: ba-secret
```

```bash
# cat /data/cosi/bucket_info.json
```

```json

>     {

Reviewer 2: nit: The json highlighting is strange. I wonder if it will be fixed if you reduce the indentation a level.
```suggestion
{
```

- First the user need to create a **COSI Bucket resource** pointing to the backend bucket.
- Then user can create BucketAccessClass and BucketAccess using the COSI Bucket CRD.
- Now the update application's credentials with BucketAccess secret, for OBC it was combination of secret and config map with keys word like AccessKey, SecretKey, Bucket, BucketHost etc. Here details in as JSON format in the [secret](#consuming-cosi-bucket).
- Finally the user need to delete the existing OBC.

### Points to remember

#### Ceph COSI Driver Requirements

- The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions
- The COSI controller should be deployed in the cluster
- Rook can manage one ceph cosi driver per Rook operator
- Rook should not modify COSI resources like Bucket, BucketAccess, BucketAccessClass, or BucketClass.

#### Rook Requirements

- Rook need to dynamically create/update the secret containing the credentials of the ceph object store for ceph COSI driver.

> - User should not be required to deploy Rook differently when using COSI and OBC for ceph object store, expect the minimal changes in the `operator.yaml` and `objectstore.yaml`.

Reviewer 2: Is the only difference in the bucket class? There are no longer changes to operator.yaml, and I don't know of any to objectstore.yaml.
```suggestion
- User should not be required to deploy Rook differently when using COSI and OBC for ceph object store.
```

# Ceph COSI Driver Support

## Targeted for v1.12

## Background

> Container Object Storage Interface (COSI) is a specification for container orchestration frameworks to manage object storage. Even though there is no standard protocol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. COSI released v1alpha1 with Kubernetes 1.25 and is not yet fully integrated by Kubernetes like CSI.

Reviewer 2: COSI is never going to be integrated the same as CSI, so maybe better to remove that sentence to avoid confusion. 
```suggestion
Container Object Storage Interface (COSI) is a specification for container orchestration frameworks to manage object storage. Even though there is no standard protocol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. COSI is released v1alpha1 with Kubernetes 1.25.
```

More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/).
It is projected that COSI will be the only supported object storage driver in the future and OBCs will be deprecated after some time.

## COSI Driver Deployment

The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. It is not deployed by Rook.

The Ceph COSI driver is deployed as a deployment with a [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver is deployed in the same namespace as the Rook operator. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:

## Integration plan with Rook

> The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There will be one COSI Driver support from Rook. The driver will be started automatically with default settings when first Object Store gets created. The driver will be stopped when last Object Store is deleted. The driver will be deployed in the same namespace as Rook operator. The user can provide additional setting via new `CephCOSIDriver` CRD which is owned by Rook.

Reviewer 2: > The driver will be stopped when last Object Store is deleted.

I would propose to follow the same pattern for uninstall as we do for the CSI driver. For CSI, an ownerReference is added to its resources that points to the Rook operator deployment. So the CSI driver is deleted whenever the Rook operator deployment is uninstalled. The same behavior seems best for the COSI driver, then Rook doesn't need any cleanup code for the COSI driver and its lifetime is the same as the operator.

### Pre-requisites

- COSI CRDs should be installed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
```

- COSI controller should be deployed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
```

### End to End Workflow

#### Changes in Common.yaml

Following contents need to be append to `common.yaml` :

- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>

#### ceph-object-cosi-controller

> The `ceph-object-cosi-controller` will start the Ceph COSI Driver pod with default settings on the Rook Operator Namespace. The controller will bring up the Ceph COSI driver when first object store is created and will stop the COSI driver when last object store is deleted only if it detect COSI Controller is running on default namespace. The controller will also watch for CephCosiDriver CRD, if defined the driver will be started with the settings provided in the CRD. If the Ceph COSI driver if up and running,it  will also create `CephObjectStoreUser` named  `cosi` for each object store which internally creates a secret rook-ceph-object-user-<objectstore-name>-cosi provides credentials for the object store. This can be specified in the BucketClass and BucketAccessClass. Also this controller ensures maximum one CephCosiDriver CRD exists in the cluster. For v1.12 supports object store belong to RGW multisite are not supported.

Reviewer 2: > For v1.12 supports object store belong to RGW multisite are not supported.

This is just saying multisite is not supported in 1.12, right? Please clarify

> The `ceph-object-cosi-controller` will start the Ceph COSI Driver pod with default settings on the Rook Operator Namespace. The controller will bring up the Ceph COSI driver when first object store is created and will stop the COSI driver when last object store is deleted only if it detect COSI Controller is running on default namespace. The controller will also watch for CephCosiDriver CRD, if defined the driver will be started with the settings provided in the CRD. If the Ceph COSI driver if up and running,it  will also create `CephObjectStoreUser` named  `cosi` for each object store which internally creates a secret rook-ceph-object-user-<objectstore-name>-cosi provides credentials for the object store. This can be specified in the BucketClass and BucketAccessClass. Also this controller ensures maximum one CephCosiDriver CRD exists in the cluster. For v1.12 the Ceph COSI Driver will be supported only for single site CephObjectStore aka object stores not configured with multisite settings like zone/zonegroup/realm.

Reviewer 2: Reposting since didn't see a response...

> The driver will be stopped when last Object Store is deleted.

I would propose to follow the same pattern for uninstall as we do for the CSI driver. For CSI, an ownerReference is added to its resources that points to the Rook operator deployment. So the CSI driver is deleted whenever the Rook operator deployment is uninstalled. The same behavior seems best for the COSI driver, then Rook doesn't need any cleanup code for the COSI driver and its lifetime is the same as the operator.

#### Transition from lib-bucket-provisioner to COSI

This applied to OBC with reclaim policy is `Retain` otherwise the bucket will be deleted when OBC is deleted. So no point in migrating the OBC with `Delete` reclaim policy.

- First the user need to create a **COSI Bucket resource** pointing to the backend bucket.
- Then user can create BucketAccessClass and BucketAccess using the COSI Bucket CRD.
- Now the update application's credentials with BucketAccess secret, for OBC it was combination of secret and config map with keys word like AccessKey, SecretKey, Bucket, BucketHost etc. Here details in as JSON format in the [secret](#consuming-cosi-bucket).
- Finally the user need to delete the existing OBC.

### Points to remember

#### Ceph COSI Driver Requirements

- The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions
- The COSI controller should be deployed in the cluster

> - Rook can manage one ceph cosi driver per Rook operator

Reviewer 3: ```suggestion
- Rook can manage one Ceph COSI driver per Rook operator
```

These terms should be consistently cased across the document.

- Rook should not modify COSI resources like Bucket, BucketAccess, BucketAccessClass, or BucketClass.

#### Rook Requirements

> - Rook need to dynamically create/update the secret containing the credentials of the ceph object store for ceph COSI driver.

Reviewer 2: Why would these creds change?

Author: RGW supports in the rotating keys, some time ago feature request with PR to supply keys for user creation. Currently, this is not possible, but in future, we may support this feature. Thats all.

> - Rook need to dynamically create/update the secret containing the credentials of the ceph object store for ceph COSI driver.

Reviewer 3: ```suggestion
- Rook needs to dynamically create/update the secret containing the credentials of the ceph object store for the ceph COSI driver.
```

> - Rook needs to dynamically create/update the secret containing the credentials of the ceph object store for the ceph COSI driver.

Reviewer 2: So this for key rotation? This is a current requirement, or a future requirement? If future, please separate the section for current or future requirements.

> - User should not be required to deploy Rook differently when using COSI and OBC for ceph object store.

Reviewer 2: Of course COSI and OBC are different. What do you mean?

> - When provisioning ceph COSI driver Rook must uniquely identify the driver name so that multiple COSI drivers or multiple Rook instances within a Kubernetes cluster will not collide.

Reviewer 2: How is this collision avoided? Is the Rook namespace prefix on the COSI driver name, similar to the CSI driver namespace prefix?

Author: That make sense

More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)
It is projected that COSI will be the only supported object storage driver in the near feature. In-tree drivers such as Ceph RGW will be replaced with their respective COSI drivers.

## Current Status of Ceph COSI Driver

The [Ceph COSI driver](https://github.com/ceph/ceph-cosi) is currently in the pre-alpha status. It is currently being tested with latest COSI Spec and APIs with images:

- cosi-controller : gcr.io/k8s-staging-sig-storage/objectstorage-controller:v20230130-v0.1.1-12-geafd6fa
- cosi-sidecar : gcr.io/k8s-staging-sig-storage/objectstorage-sidecar/objectstorage-sidecar:v20230130-v0.1.0-24-gc0cf995

The Ceph COSI driver supports RGW with ceph storage as backend with s3 protocol. Later on, it will be extended to support other protocols such as Swift and different backends for RGW via Zipper.

## COSI Driver Deployment

> The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:

Reviewer 2: > The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. 

It really has to be in the default namespace? If it had to be a specific a specific namespace, I'm surprised it's not `kube-system`.

> The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). 

What state does the driver have? Why does it need to be a statefulset? Would it be backed by an rbd volume? 

> The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:

Let's expand on this more. Based on our conversation, sounds like we would put it in the same namespace as the CephObjectStore CR, unless we decide there is a need to put it with the operator.

Author: > > The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace.
> 
> It really has to be in the default namespace? If it had to be a specific a specific namespace, I'm surprised it's not `kube-system`.

Atleast they resources are defined like setting up in default namespace https://github.com/kubernetes-sigs/container-object-storage-interface-controller/tree/master/resources
> 
> > The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar).
> 
> What state does the driver have? Why does it need to be a statefulset? Would it be backed by an rbd volume?
> 

Currently, it is stateless, it does not have data stored internally atm, not sure whether we need storage. 

> > The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
> 
> Let's expand on this more. Based on our conversation, sounds like we would put it in the same namespace as the CephObjectStore CR, unless we decide there is a need to put it with the operator.

Sure

Reviewer: It seems as though we should...
1. Figure out if the COSI controller needs installed in the default namespace. It seems to me that the namespace should be able to be arbitrary unless there is some unnecessary coding assumption in the COSI controller. This doc should only need to be detailed enough to explain that the COSI controller is a necessary prerequisite with a link to COSI docs about the controller itself.
2. Change the Ceph-COSI driver to be a deployment since we don't need a statefulset currently.

-----

I also see some implicit decision-making happening here that I would prefer to have a record of. Technical designs shouldn't leave any decision as implicit; we want to have a reason for each one, and the easiest way to do that is to actually discuss each option that is available to us.

I also think we include in the doc a pro/con list of the different namespaces to install the driver in.
1. The rook system namespace such that it operates for multiple ceph clusters
2. The cluster namespace such that the driver is run once for each ceph cluster

We should also include a pro/con list of how many COSI drivers should be run:
1. Just one in the whole k8s system
2. One for every rook operator
4. One for every CephCluster
5. One for every CephObjectStore

These decisions aren't necessarily just for Rook. These decisions also affect the design of the Ceph COSI controller itself. 

-----

As I think about it more for myself, I think the best decision would be to install one Ceph COSI driver (with the option of multiple replicas) in the whole system. Rook's CephObjectStore controller should then create the necessary resources to allow the COSI driver to allow bucket provisioning for said CephObjectStore.

This decision does not assume that the decisions made for Ceph-CSI are the best or correct ones.

For the pros, I believe that this more closely matches the design paradigms of Kubernetes and other CNCF projects:
- Similarly to the k8s API server, there is only one, and it can scale to meet demand
- Because there is only one, cloud install environments (where CPU and memory demands cost ever more money) don't need to worry about excess resource requirements due to needing more instances than necessary
- The Ceph-COSI driver itself is an independent component from Rook
  - This allows other components to integrate with Ceph-COSI if desired without needing Rook 
  - This means that Rook won't affect the ability of other components to integrate with Ceph-COSI
  - Rook can use Ceph-COSI's defined APIs (whether CRDs or via Secrets/ConfigMaps) to integrate itself as desired
- Rook's codebase does not need to codify or document how to install or configure Ceph-COSI. The responsibility for that is moved to the ceph/ceph-cosi project itself which reduces duplication of effort and reduces the possibility of copy-paste errors between the projects which we sometimes see with Ceph-CSI. It allows Ceph-CSI to make necessary changes (especially to its deployment) more quickly without need for changes to the Rook codebase.

The downsides are that:
- Rook doesn't control Ceph-COSI, and so it might be modified or removed such that Rook's integration wouldn't work
  - Having a well-defined and backwards-compatible Ceph-COSI API will help with this
  - This is in some way the nature of everything in the CNCF ecosystem. We could recommend that users put a finalizer on the Ceph-COSI deployment when Rook is installed stating Rook's intended and ongoing usage of it, which helps mitigate this issue (and is one reason why finalizers exist)

-----

On that note, we should have a design review once the alpha COSI controller is done. I have seen our group often make the mistake of assuming that an initially-working piece of software is a good foundation to build off of for the future, when that is most definitely not the case in any of my experience.

In particular, I want to push to make sure the COSI driver is using controller-runtime and is making full use of structured logging. I would also suggest having a versioned, backwards/forwards-compatible API for Ceph-COSI's configs that can be used by Rook.

@thotz, when Ceph-COSI alpha is done or mostly done, let's organize a design review with the important stakeholders. That would be me and Travis from Rook; Ali who also bridges the RGW-Rook gap; and this could also include Matt B. and/or Casey B. on the RGW side.

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: objectstorage-provisioner-role
  labels:
    app.kubernetes.io/part-of: container-object-storage-interface
    app.kubernetes.io/component: driver-ceph
    app.kubernetes.io/version: main

>     app.kubernetes.io/name: cosi-driver-ceph

Reviewer: Let's allow Ceph-COSI to define these.

rules:
- apiGroups: ["objectstorage.k8s.io"]
  resources: ["buckets", "bucketaccesses", "bucketclaims", "bucketaccessclasses", "buckets/status", "bucketaccesses/status", "bucketclaims/status", "bucketaccessclasses/status"]
  verbs: ["get", "list", "watch", "update", "create", "delete"]
- apiGroups: ["coordination.k8s.io"]
  resources: ["leases"]
  verbs: ["get", "watch", "list", "delete", "update", "create"]
- apiGroups: [""]
  resources: ["secrets", "events"]

>   verbs: ["get", "delete", "update", "create"]

Reviewer: Can we strip this down further? Especially if Ceph-COSI is going to look for RGW secrets in any namespace, limiting the access to specifically-named secrets follows the principles of least privilege.

```

> All the COSI CRDs are installed from <http://github.com/kubernetes-sigs/container-object-storage-interface-api>

Reviewer 2: Are there deployment instructions? I don't see on that page how it's deployed. It would be helpful to know what steps are required as a prereq for Rook to install the Ceph COSI driver.

Author: I will add Prereq part in the design doc

## Integration plan with Rook

> The aim to support alpha version of COSI in Rook v1.12 and depending on development of COSI, it will be extended to beta and GA versions in Rook Operaotr. There should be option in `Operator.yaml` to bring up the COSI controller and another option at object store level to bring up the ceph COSI driver.

Reviewer 2: I'm envisioning it as a separate step to start the COSI controller, instead of Rook doing it with an operator option, similar to how we document [installing Prometheus](https://rook.io/docs/rook/latest/Storage-Configuration/Monitoring/ceph-monitoring/#prometheus-operator) as a prereq for monitoring.

> The aim to support alpha version of COSI in Rook v1.12 and depending on development of COSI, it will be extended to beta and GA versions in Rook Operaotr. There should be option in `Operator.yaml` to bring up the COSI controller and another option at object store level to bring up the ceph COSI driver.

Reviewer: Do you have the option of installing a spell-checker in your editor?

```suggestion
The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to bring up the COSI controller and another option at object store level to bring up the Ceph COSI driver.
```

### How Rook can improve COSI driver reliability

> Rook can ensure that resources need for ceph cosi driver such as ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for  multiple ceph object stores in the cluster.

Reviewer 2: > Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions

Rook doesn't create RBAC unless it's with the helm chart, or as a separate yaml. I would think this would be done with the prereq steps if possible.

Reviewer 2: > Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster.

Rook already prevents deletion of the object store if there are any buckets.

Reviewer 2: > Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver.

How about a separate section in the design doc for the details about the secret?

Author: > > Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster.
> 
> Rook already prevents deletion of the object store if there are any buckets.

Yes I know, just listed here to make sure we are covering that

> Rook can ensure that resources need for ceph cosi driver such as ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for  multiple ceph object stores in the cluster.

Reviewer: I'm not seeing consistent capitalization or usage of naming. In general, "Ceph", "Rook", and "COSI" should be capitalized unless referring to a specific process/container/pod. I would also suggest eliminating any possibility for unclear intent by using `CephCluster` or `CephObjectStore` when that resource is being discussed; similarly Ceph RGW when that is relevant.

```suggestion
Rook can ensure that resources need for Ceph COSI driver such as Ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for  multiple ceph object stores in the cluster.
```

### Coexistence of COSI and libbucket provisioner

Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can co exist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The libbucket probvisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.

### Ceph COSI Driver Requirements

- Ceph Object Store should be deployed and running in the cluster
- The credentials/endpoint for the ceph object store should be available by creating ceph object store user with proper permissions
- The COSI controller should be deployed in the cluster
- Rook can able to manage multiple ceph cosi drivers
- Rook should not modify cosi resource like Bucket, BucketAccess, BucketClaim, BucketAccessClass etc.

### Rook Requirements

- Rook must include all the RBAC permissions required by the COSI for the deployment.

> - Rook need to install all the COSI CRDs.

Reviewer 2: The CRDs must be a prereq since rook doesn't own them.

- Rook need to dynamically create/update the secret containing the credentials of the ceph object store for ceph COSI driver.

> - User should not be required to deploy Rook differently when using COSI and OBC for ceph object store, expect the minimal changes in the `Operator.yaml` and `ObjectStore.yaml`.

Reviewer 2: I hope we don't need changes in operator.yaml. I'm wondering if we would only need a simple setting in the CephObjectStore CR such as:
```yaml
  cosiDriver:
    enabled: true
```

> - User should not be required to deploy Rook differently when using COSI and OBC for ceph object store, expect the minimal changes in the `Operator.yaml` and `ObjectStore.yaml`.

Reviewer: Why are the names of files, which are case sensitive, arbitrarily capitalized? Does this refer to different files?

```suggestion
- User should not be required to deploy Rook differently when using COSI and OBC for ceph object store, expect the minimal changes in the `operator.yaml` and `objectstore.yaml`.
```

> - When provisioning ceph COSI driver Rook must uniquely identify the driver/provisioner name so that multiple COSI drivers or multiple Rook instances within a (Kubernetes) cluster will not collide.

Reviewer: Is it "driver" or "provisioner"? Let's pick one term and stick to it for the entire doc.

```suggestion
- When provisioning ceph COSI driver Rook must uniquely identify the driver/provisioner name so that multiple COSI drivers or multiple Rook instances within a (Kubernetes) cluster will not collide.
```

Author: missed this one, will address in next update

> - When provisioning ceph COSI driver Rook must uniquely identify the driver/provisioner name so that multiple COSI drivers or multiple Rook instances within a (Kubernetes) cluster will not collide.

Reviewer: Parens around "(Kubernetes)" are unnecessary and IMO confusing.

```suggestion
- When provisioning ceph COSI driver Rook must uniquely identify the driver/provisioner name so that multiple COSI drivers or multiple Rook instances within a Kubernetes cluster will not collide.
```

# Ceph COSI Driver Support

## Targeted for v1.12

## Background

Container Object Storage Interface (COSI) is a specification for container orchestration frameworks to manage object storage. Even though there is no standard procotol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. COSI released v1alpha1 with Kubernetes 1.25. COSI is a new project and is not yet fully integrated by Kubernetes.
More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)
It is projected that COSI will be the only supported object storage driver in the near future. In-tree drivers such as Ceph RGW will be replaced with their respective COSI drivers.

## Current Status of Ceph COSI Driver

> The [Ceph COSI driver](https://github.com/ceph/ceph-cosi) is currently in the pre-alpha status. It is currently being tested with latest COSI Spec and APIs with images:

Reviewer 2: The focus of the design doc is how we want the feature to work, I'm not sure we need to mention the current state of the pre-alpha driver.

Author: Sure will remove this para

> - cosi-controller : gcr.io/k8s-staging-sig-storage/objectstorage-controller:v20230130-v0.1.1-12-geafd6fa

Reviewer 2: There are newer images now, right? Are there also more official version release tags?

Author: no official tags as of now

The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:

## Integration plan with Rook

The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.

### Pre-requisites

- COSI CRDs should be installed in the cluster via following command

```bash
kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
```

- COSI controller should be deployed in the cluster via following command

```bash

> kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller

Reviewer 2: How does this COSI controller relate to the controller enabled in the rook operator.yaml below?

Author: It just validates whether cosi controller running or not.

```

- Create Service Account and RBAC for COSI driver

```bash
kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml
kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml
```

### Changes in Operator.yaml

> Below option validates whether the COSI controller is deployed in the cluster. If it is not deployed, it will log an warning message and when user tries to create object store woth cosi, it will reconcile until the controller is deployed.

Reviewer 2: See the codespell CI issues

```

>   cosiController:

Reviewer 2: operator.yaml settings are all env vars as upper case. Is this an example from the helm chart?

    enabled: true
```

> ### Changes in ObjectStore.yaml

Reviewer 2: This section is for the CephObjectStore CR?
```suggestion
### Changes in CephObjectStore
```

Following option will bring up the Ceph COSI driver for the RGW server. It will also create `CephObjectStoreUser`  objectstorage-provisioner which provides credentials for Ceph COSI driver.
```
  cephCosiDrvierSpec:
    enabled: true
```

### How Rook can improve COSI driver reliability

> Rook can ensure that resources need for Ceph COSI driver such as Ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster which already present with current codebase. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for multiple ceph object stores in the cluster. Rook also need to support update of COSI-Controller and COSI-Driver images independently

Author: I will have more idea when all the content is finalised and will address accordingly

> Rook can ensure that resources need for Ceph COSI driver such as Ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster which already present with current codebase. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for multiple ceph object stores in the cluster. Rook also need to support update of COSI-Controller and COSI-Driver images independently

Reviewer 2: Let's have a design discussion to finalize as much as we can. We really don't have much more time to answer design questions if we are to include this in the 1.12 release.

### Coexistence of COSI and libbucket provisioner

> Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can co exist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The libbucket probvisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.

Reviewer 2: Why would we want to support the same backend bucket to be managed from COSI and OBC? Will that help us transition to COSI? If so, how about a new section that talks about transitioning from OBC to COSI?

### Ceph COSI Driver Requirements

- Ceph Object Store should be deployed and running in the cluster
- The credentials/endpoint for the ceph object store should be available by creating ceph object store user with proper permissions
- The COSI controller should be deployed in the cluster

> - Rook can able to manage multiple ceph cosi drivers

Reviewer 2: I'm still not clear... Why do we need multiple COSI drivers? Or can we just support one COSI driver, just like we do for CSI?

Author: Yes thats based on last discussion. Currently, Ceph COSI driver only supports one endpoint, We can have one single driver for every ceph object stores in Rook. Since ceph cosi driver is light weight we thought of having multiple drivers than a single one. But we can change what ever makes good scalable design

Reviewer 2: Could you add a section that lists which resources/components/pods will be created by Rook (or not included with Rook)? It will help understand the single COSI controller, the object-store-specific COSI driver, etc.

Reviewer: The latest discussion was that there should be one Ceph COSI driver per k8s cluster. Let's adjust this design to reflect that.
